R/knitr-engine.R

```diff
@@ -0,0 +1,160 @@
+  # best to emulate the return format of 'evaluate::evaluate()'
+  code <- options$code
+  n <- length(code)
+  if (n == 0)
+    return(list())
+  
+  # use 'ast.parse()' to parse Python code and collect line numbers, so we
+  # can split source code into statements
+  pasted <- paste(code, collapse = "\n")
+  parsed <- ast$parse(pasted, "<string>")
+  
+  # iterate over top-level nodes and extract line numbers
+  lines <- vapply(parsed$body, function(node) {
+    node$lineno
+  }, integer(1))
+  
+  # convert from lines to ranges
+  starts <- lines
+  ends <- c(lines[-1] - 1, length(code))
+  ranges <- mapply(c, starts, ends, SIMPLIFY = FALSE)
+  
+  # line index from which source should be emitted
+  pending_source_index <- 1
+  
+  # plots captured from Python which need to be emitted
+  pending_plots <- list()
+  
+  # actual outputs to be returned to knitr
+  outputs <- list()
+  
+  for (range in ranges) {
+    
+    # evaluate current chunk of code
+    snippet <- extract(code, range)
+    captured <- py_capture_output(
+      py_run_string(snippet, convert = FALSE)
+    )
+    
+    if (nzchar(captured) || length(pending_plots)) {
+      
+      # append pending source to outputs
+      outputs[[length(outputs) + 1]] <- structure(
+        list(src = extract(code, c(pending_source_index, range[2]))),
+        class = "source"
+      )
+      
+      # append captured outputs
+      if (nzchar(captured))
+        outputs[[length(outputs) + 1]] <- captured
+      
+      # append captured images / figures
+      if (length(pending_plots)) {
+        for (plot in pending_plots)
+          outputs[[length(outputs) + 1]] <- plot
+        pending_plots <- list()
+      }
+      
+      # update pending source range
+      pending_source_index <- range[2] + 1
+    }
+  }
+  
+  # if we have leftover input, add that now
+  if (pending_source_index < n) {
+    leftover <- extract(code, c(pending_source_index, n))
+    outputs[[length(outputs) + 1]] <- structure(
+      list(src = leftover),
+      class = "source"
+    )
+  }
+  
+  # TODO: development version of knitr supplies new 'engine_output()'
+  # interface -- use that when it's on CRAN
+  # https://github.com/yihui/knitr/commit/71bfd8796d485ed7bb9db0920acdf02464b3df9a
+  wrap <- yoink("knitr", "wrap")
+  wrap(outputs, options)
+  
+}
+
+eng_python_initialize <- function(options, context, envir) {
+  
+  if (is.character(options$engine.path))
+    use_python(options$engine.path[[1]])
+  
+  eng_python_initialize_matplotlib(options, context, envir)
+}
+
+eng_python_initialize_matplotlib <- function(options,
+                                             context,
+                                             envir)
+{
+  if (!py_module_available("matplotlib"))
+    return()
+  
+  matplotlib <- import("matplotlib", convert = FALSE)
+  plt <- matplotlib$pyplot
+  
+  # rudely steal 'plot_counter' (used below), and reset
+  # it when we're done
+  plot_counter <- yoink("knitr", "plot_counter")
```

Comment:
I don't quite understand the `defer()` magic below, but the counter should be reset in `eng_python()` after `wrap()` is done (e.g. in `on.exit()`), otherwise I guess you will always get the same figure file path `chunk-label-1.png`.

Reply:
`defer()` is basically a way of calling `on.exit()` in an arbitrary frame; ie, I use the fact that this function will only be called within `eng_python()`, and attach the `on.exit()` handler there.

Reply:
Okay, that sounds clever!

Reply:
FWIW defer has been in the CRAN version of withr for a while now as `withr::defer()` if you don't want to duplicate the code here (but I guess presumably you do?).

Reply:
It's so great to have `withr::defer()`. Solved my headache elegantly. Thanks.